### ISSUE TITLE
Uncaught Error: useLocation() may be used only in the context of a <Router> component.

### ISSUE THREAD

**Issue by BoyPersoon** [react-auth-kit/react-auth-kit#872]
Unfortunately I am getting the following error in my App, when implementing the basic example from the docs.

The code below results in the following error:
```
Uncaught Error: useLocation() may be used only in the context of a <Router> component.
```

code:

```
const Router = () => (
    <AuthProvider
        authType={'cookie'}
        authName={'_auth'}
        cookieDomain={window.location.hostname}
        cookieSecure={window.location.protocol === 'https:'}
    >
        <BrowserRouter>
            <Routes>
                <Route
                    path={'/'}
                    element={
                        <RequireAuth loginPath={'/login'}>
                            <div>Secure</div>
                        </RequireAuth>
                    }
                />
            </Routes>
        </BrowserRouter>
    </AuthProvider>
);
```

package.json
```
    "react-router": "^6.2.1",
    "react-router-dom": "6",
    "react-auth-kit": "^2.0.4",
```

**Expected behavior**
This simple example should not result in the listed error

**Desktop (please complete the following information):**
 - Browser 
 Chrome + Firefox

**Comment by darkmatter18** [comment]
I'll check the issue and will let you know.

**Comment by izaktan** [comment]
I have the same problem

**Comment by izaktan** [comment]
> @izaktan Are you wrapping the routes inside any browser element like `BrowserRouter`?

I am not wrapping that directly. There are additional layers. I got something like this:
  ```
<AuthProvider
    authName={"_auth"}
    authType={"cookie"}
    cookieDomain={window.location.hostname}
    cookieSecure={window.location.protocol === "https:"}
  >
    <BrowserRouter>
        <AppPermissionControllerProvider>
            <ThemeProvider>
                <Routes />
            </ThemeProvider>
        </AppPermissionControllerProvider>
    </BrowserRouter>
  </AuthProvider>
```

**Comment by izaktan** [comment]
So why? In the document it says `AuthProvider should wrap the BrowserRouter or HashRouter, otherwise PrivateRoute will not work and throw an error.`

**Comment by Kyle-Gagnon-99** [comment]
Hey! I was wondering if there was an update on this? I did try wrapping the ```BrowserRouter``` around ```AuthProvider``` but I get the following error:

```You cannot render a <Router> inside another <Router>. You should never have more than one in your app.```

Now I did pick up ReactJS within the last two weeks but I am not sure what is going on. Here is some code snippit from me:

```
    <BrowserRouter>
      <AuthProvider authType={"localstorage"} authName={"_auth"}>
        <Routes>
          <Route element={<Main />} exact path="/" />
          <Route element={<ContactUs />} exact path="/contactus" />
          <Route element={<Login />} exact path="/login" />
          <Route
            element={<RequireAuth loginPath="/login">
              <HelloWorld />
            </RequireAuth>}
            exact
            path="/helloworld"
          />
        </Routes>
      </AuthProvider>
    </BrowserRouter>
```

**Comment by faner11** [comment]
This problem occurs in react 17, and 18 is normal

**Comment by Clayton-Noyes** [comment]
I am still getting this issue in 2023. Currently my react and react-auth-kit versions are the following:
```
    "react": "^18.2.0",
    "react-auth-kit": "^2.12.2",
```

The contents of my App.tsx file is the following:
```
// Built in React Components
import { FC, useMemo } from "react";
import { BrowserRouter, Routes, Route } from "react-router-dom";

// Third Party Libraries
import { CssBaseline } from "@mui/material";
import { ThemeProvider } from "@emotion/react";
import { AuthProvider, RequireAuth } from "react-auth-kit";

// App-specific Pages and components
import buildTheme from "./styles/theme";
import LoginView from "./views/Login";
import SignUp from "./views/SignUp";
import Dashboard from "./views/Dashboard/";
import NoMatch from "./views/NoMatch";

const App: FC = () => {
  // Handle Theming
  // const prefersDarkMode = useMediaQuery("(prefers-color-scheme: dark)");
  const mode = "light";
  // const mode = "dark"; // UNCOMMENT to work in dark mode, but do not commit while line is active
  const theme = useMemo(() => buildTheme(mode), [mode]);

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline enableColorScheme>
        <AuthProvider
          authType={"cookie"}
          authName={"_auth"}
          cookieDomain={window.location.hostname}
          cookieSecure={window.location.protocol === "https:"}
        >
          <BrowserRouter>
            <Routes>
              {/* Unauthorized People can access these routes */}
              <Route path="/" element={<LoginView />} />
              <Route path="/login" element={<LoginView />} />
              <Route path="/signup" element={<SignUp />} />

              {/* Authorized People can access these routes */}

              <Route
                path="/dashboard"
                element={
                  <RequireAuth loginPath="/login">
                    <Dashboard />
                  </RequireAuth>
                }
              />

              {/* Unable to find a page with this route */}
              {/* <Route path="*" element={<NoMatch />} /> */}
            </Routes>
          </BrowserRouter>
        </AuthProvider>
      </CssBaseline>
    </ThemeProvider>
  );
};

export default App;
```

The error I get is `useLocation() may be used only in the context of a <Router> component.` I do not think this issue should be marked as stale, nor closed as it appears that there is still an issue with the `RequireAuth` component. Which is the culprit according to the stack trace output in the error message:
```
useLocation() may be used only in the context of a <Router> component.
    at invariant (http://localhost:3000/static/js/bundle.js:60293:11)
    at useLocation (http://localhost:3000/static/js/bundle.js:64064:102)
    at RequireAuth (http://localhost:3000/static/js/bundle.js:58478:79)
    at renderWithHooks (http://localhost:3000/static/js/bundle.js:79352:22)
    at mountIndeterminateComponent (http://localhost:3000/static/js/bundle.js:82638:17)
    at beginWork (http://localhost:3000/static/js/bundle.js:83934:20)
    at HTMLUnknownElement.callCallback (http://localhost:3000/static/js/bundle.js:68944:18)
    at Object.invokeGuardedCallbackDev (http://localhost:3000/static/js/bundle.js:68988:20)
    at invokeGuardedCallback (http://localhost:3000/static/js/bundle.js:69045:35)
    at beginWork$1 (http://localhost:3000/static/js/bundle.js:88919:11)
```

**Comment by Clayton-Noyes** [comment]
After messing around with trying to get this to work I finally tried resetting my node_modules that my app was depending on. I am using `yarn` to manage dependencies.

I followed these steps:
1. `yarn remove react-router-dom react-auth-kit` to remove them from my dependencies
2. deleted my node_modules folder
3. checked to make sure that my `package.json` file was purged of mentions of `react-router-dom` and `react-auth-kit` along with their type declarations
4. reinstalled my node_modules via `yarn install`
5. then proceeded to install both react-router-dom and react-auth-kit via: `yarn install react-auth-kit react-router-dom`

I was instantly able to see my protected route and the error was no longer being thrown. Looking at my package.json file it shows that react-auth-kit was installed to version 2.12.2 and that react-router-dom was actually behind on its version. It went from `"react-router-dom": "^6.10.0"` to `"react-router-dom": "^6.11.1". There must have been some change between 6.10 to 6.11 that inadvertently fixed this.

So, if you are reading this now and you are encountering this useLocation() error with your application, even though you have followed the tutorial posted on react-auth-kit's website and documentation; then chances are the issue is with your dependencies. I would follow the above procedure to ensure that you install the most up-to-date version of react-router-dom and react-auth-kit, after purging your node_modules folder. This fixed the error for me.

React Auth Kit team: thanks for building a package that handles authentication and session management so effortlessly. I really appreciate your efforts in building this.
